### PR TITLE
adds a debouncing decorator function

### PR DESCRIPTION
**Pull request by Dan1001**
This is a simple way of preventing a rule from firing multiple times when its trigger fires multiple times in quick succession.
Neater and simpler than adding a lock or timer into each function manually.

I am, however, a python novice, and have no doubt this can be improved upon!

### CODE REVIEW

File: Core/automation/lib/python/core/utils.py [hunk 1]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+
+@rule("Main test function")
+@when("Item TestSwitch received command ON") 
+@debounce
+def main_test(event):
+
+   log.info("Test switch on")
+   
+"""
+    def wrapper_debounce(event):
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:319]: Have you tested this much and does logging work properly? I'd think you'd need to use [functools.wraps](https://docs.python.org/2/library/functools.html#functools.wraps), or things may get messed up in the other decorators.

File: Core/automation/lib/python/core/utils.py [hunk 2]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+        global BouncingRuleTracker
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:320]: There's no need to use global for a collection.

File: Core/automation/lib/python/core/utils.py [hunk 3]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+        min_delay = 2
+
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:322]: Remove empty lines

File: Core/automation/lib/python/core/utils.py [hunk 4]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
         uid = "{}_{}".format("jython", uid)
     uid = re.sub(r"__+", "_", uid)
     return uid
+
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:282]: Remove empty line.

File: Core/automation/lib/python/core/utils.py [hunk 5]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+   
+"""
+    def wrapper_debounce(event):
+        global BouncingRuleTracker
+        min_delay = 2
+
+        current_rule = str(func.__name__)
+
+        report = "Started rule " + current_rule + \
+            ", triggered by " + str(event.itemName)
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:326]:
```suggestion
report = u"debounce: started rule '{}', triggered by '{}'".format(current_rule, event.itemName)
```

File: Core/automation/lib/python/core/utils.py [hunk 6]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+
+        if current_rule in BouncingRuleTracker:
+
+            if BouncingRuleTracker[current_rule]["running"] == True:
+                # Note currently openhab seems to always run rules sequentially
+                # so this condition will never be satisfied - but still
+                # seems sensible to deal with this case
+                log.info(report + ", which is already running - so exiting now")
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:334]:
```suggestion
                log.debug(u"{}, which is already running - so exiting now.".format(report))
```

File: Core/automation/lib/python/core/utils.py [hunk 7]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+                return
+
+            actualdelay = round(
+                (time.time() - BouncingRuleTracker[current_rule]["timetrack"]), 1)
+            report = report + ". Last finished " + \
+                str(actualdelay) + " seconds ago"
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:340]:
```suggestion
            report = u"{}. Last finished {} seconds ago.".format(report, actualdelay)
```

File: Core/automation/lib/python/core/utils.py [hunk 8]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+
+            if actualdelay < min_delay:
+                report = report + ", so debouncing and not running rule"
+                log.info(report)
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:344]:
```suggestion
                log.debug(u"{}, so debouncing and not running rule".format(report))
```

File: Core/automation/lib/python/core/utils.py [hunk 9]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+                return
+
+        else:
+            report = report + ". This is the first time the rule has run."
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:348]:
```suggestion
            report = u"{}. This is the first time the rule has run.".format(report)
```

File: Core/automation/lib/python/core/utils.py [hunk 10]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+
+        log.debug(report)
+
+        func(event)
+
+        ruletime = round(
+            (time.time() - BouncingRuleTracker[current_rule]["timetrack"]), 1)
+
+        log.debug("Finishing rule " + current_rule + " triggered by " +
+                 str(event.itemName) + ". Rule took " + str(ruletime) + " seconds to run.")
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:362]:
```suggestion
        log.debug(u"Finishing rule '{}', triggered by '{}'. Rule took {} seconds to run.".format(current_rule, event.itemName, ruletime))
```

File: Core/automation/lib/python/core/utils.py [hunk 11]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
...
+            BouncingRuleTracker[current_rule] = {}
+
+        BouncingRuleTracker[current_rule]["timetrack"] = time.time()
+        BouncingRuleTracker[current_rule]["running"] = True
+
+        log.debug(report)
+
+        func(event)
+
+        ruletime = round(
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:358]: Use snake_case.

File: Core/automation/lib/python/core/utils.py [hunk 12]
```diff
@@ -279,3 +279,90 @@ def validate_uid(uid):
         uid = "{}_{}".format("jython", uid)
     uid = re.sub(r"__+", "_", uid)
     return uid
+
+
+BouncingRuleTracker = {}
+
+
+def debounce(func):
```
Comment (openhab-5iver) [inline Core/automation/lib/python/core/utils.py:287]: Truncating the parameter only makes things less readable. Use function.

**Comment by Dan1001** [comment]
Thank you!

This is all a bit beyond me - so if you can submit that would be great. I don't want to use my name, but will send you a PM with my details.

Also, I'm not sure how to amend the code now, but if you see the forum thread, I've amended the function to prevent the unintended feature of blocking a rule running again if it ever fails with an error.

**Pull request closed without merging**